lib/src/conflicts.rs

```diff
@@ -273,37 +287,73 @@ fn write_conflict_marker(
     }
 }
 
-/// Parse a conflict marker from a line of a file. The conflict marker must have
-/// the correct length (CONFLICT_MARKER_LEN).
-fn parse_conflict_marker(line: &[u8]) -> Option<ConflictMarkerKind> {
+/// Parse a conflict marker from a line of a file. The conflict marker may have
+/// any length (even less than MIN_CONFLICT_MARKER_LEN).
+fn parse_conflict_marker_any_len(line: &[u8]) -> Option<ConflictMarker> {
     let first_byte = *line.first()?;
     let kind = ConflictMarkerKind::parse_byte(first_byte)?;
     let len = line.iter().take_while(|&&b| b == first_byte).count();
 
-    if len != CONFLICT_MARKER_LEN {
-        return None;
-    }
-
     if let Some(next_byte) = line.get(len) {
         // If there is a character after the marker, it must be ASCII whitespace
         if !next_byte.is_ascii_whitespace() {
             return None;
         }
     }
 
-    Some(kind)
+    Some(ConflictMarker { kind, len })
+}
+
+/// Parse a conflict marker, expecting it to be at least a certain length. Any
+/// shorter conflict markers are ignored.
+fn parse_conflict_marker(line: &[u8], expected_len: usize) -> Option<ConflictMarkerKind> {
+    parse_conflict_marker_any_len(line)
+        .filter(|marker| marker.len >= expected_len)
+        .map(|marker| marker.kind)
+}
+
+/// Given a Merge of files, choose the conflict marker length to use when
+/// materializing conflicts.
+pub fn choose_materialized_conflict_marker_len<T: AsRef<[u8]>>(single_hunk: &Merge<T>) -> usize {
+    let max_existing_marker_len = single_hunk
+        .iter()
+        .flat_map(|file| file.as_ref().lines_with_terminator())
+        .filter_map(parse_conflict_marker_any_len)
+        .map(|marker| marker.len)
+        .max()
+        .unwrap_or_default();
+
+    max_existing_marker_len
+        .saturating_add(CONFLICT_MARKER_LEN_INCREMENT)
+        .max(MIN_CONFLICT_MARKER_LEN)
 }
 
 pub fn materialize_merge_result<T: AsRef<[u8]>>(
     single_hunk: &Merge<T>,
     conflict_marker_style: ConflictMarkerStyle,
     output: &mut dyn Write,
+) -> io::Result<()> {
+    let conflict_marker_len = choose_materialized_conflict_marker_len(single_hunk);
+
+    materialize_merge_result_with_marker_len(
```

Comment:
nit: perhaps `match &merge._result { .. }` can be inlined here like other `materialize_*` functions?